File: backend/src/modules/ledger/documents.ts
import { and, eq } from "drizzle-orm";
import { db, type DbTransaction } from "../../db/index.js";
import {
  accounts,
  documentSequences,
  documents,
  type documentTypeEnum,
} from "../../db/schema.js";
import { ApiError } from "../../lib/errors.js";
import {
  createInvoice,
  createLoanIn,
  createLoanPayment,
  createReceipt,
} from "../arap/service.js";
import { postJournal, type JournalLineInput } from "./journal.js";

export type DocumentKind =
  | "cash_in"
  | "cash_out"
  | "transfer"
  | "capital"
  | "loan_in"
  | "loan_payment"
  | "invoice"
  | "receipt";

type DocumentType = (typeof documentTypeEnum.enumValues)[number];

const FULLY_IMPLEMENTED_KINDS = new Set<DocumentKind>([
  "cash_in",
  "cash_out",
  "transfer",
  "capital",
]);

const KIND_TO_DOC_TYPE: Record<DocumentKind, DocumentType> = {
  cash_in: "receipt",
  cash_out: "payment",
  transfer: "journal",
  capital: "journal",
  loan_in: "journal",
  loan_payment: "payment",
  invoice: "invoice",
  receipt: "receipt",
};

const DEFAULT_PREFIX: Record<DocumentType, string> = {
  invoice: "INV-",
  bill: "BIL-",
  payment: "PY-",
  receipt: "RC-",
  journal: "JV-",
  adjustment: "ADJ-",
  other: "DOC-",
};

export interface CreateDocumentInput {
  orgId: string;
  userId: string;
  kind: DocumentKind;
  amount: number;
  memo?: string;
  cashAccountId?: string;
  counterAccountId?: string;
  isPrive?: boolean;
  date?: string;
  dueDate?: string;
  contactId?: string;
  openItemId?: string;
  documentId?: string;
}

function parseAmount(amount: number): number {
  const rounded = Math.round(amount);
  if (!Number.isFinite(rounded) || rounded <= 0) {
    throw new ApiError("INVALID_AMOUNT", "Amount must be a positive integer", 400);
  }
  return rounded;
}

function parseDate(date?: string): Date {
  const value = date ?? new Date().toISOString().slice(0, 10);
  if (!/^\d{4}-\d{2}-\d{2}$/.test(value)) {
    throw new ApiError("INVALID_DATE", "Date must be YYYY-MM-DD", 400);
  }
  return new Date(`${value}T00:00:00.000Z`);
}

async function getAccount(
  orgId: string,
  accountId: string,
  label: string,
): Promise<{ id: string; type: string; isCash: boolean; code: string }> {
  const [account] = await db
    .select({
      id: accounts.id,
      type: accounts.type,
      isCash: accounts.isCash,
      code: accounts.code,
    })
    .from(accounts)
    .where(and(eq(accounts.id, accountId), eq(accounts.orgId, orgId)))
    .limit(1);

  if (!account) {
    throw new ApiError("ACCOUNT_NOT_FOUND", `${label} account not found`, 404);
  }

  return account;
}

async function getDefaultCashAccount(orgId: string): Promise<string> {
  const [cash] = await db
    .select({ id: accounts.id })
    .from(accounts)
    .where(and(eq(accounts.orgId, orgId), eq(accounts.isCash, true)))
    .limit(1);

  if (!cash) {
    throw new ApiError("ACCOUNT_NOT_FOUND", "No cash account configured", 404);
  }

  return cash.id;
}

async function getSystemAccountByCode(
  orgId: string,
  code: string,
): Promise<string> {
  const [account] = await db
    .select({ id: accounts.id })
    .from(accounts)
    .where(and(eq(accounts.orgId, orgId), eq(accounts.code, code)))
    .limit(1);

  if (!account) {
    throw new ApiError(
      "ACCOUNT_NOT_FOUND",
      `System account ${code} not found`,
      404,
    );
  }

  return account.id;
}

async function nextDocumentNumber(
  orgId: string,
  documentType: DocumentType,
  tx: DbTransaction,
): Promise<string> {
  const [existing] = await tx
    .select()
    .from(documentSequences)
    .where(
      and(
        eq(documentSequences.orgId, orgId),
        eq(documentSequences.documentType, documentType),
      ),
    )
    .limit(1);

  let seq = existing;

  if (!seq) {
    const [created] = await tx
      .insert(documentSequences)
      .values({
        orgId,
        documentType,
        prefix: DEFAULT_PREFIX[documentType],
        nextNumber: 1,
        padding: 4,
      })
      .returning();
    seq = created!;
  }

  const number = `${seq.prefix}${String(seq.nextNumber).padStart(seq.padding, "0")}`;

  await tx
    .update(documentSequences)
    .set({ nextNumber: seq.nextNumber + 1, updatedAt: new Date() })
    .where(eq(documentSequences.id, seq.id));

  return number;
}

function buildJournalLines(input: CreateDocumentInput & { amount: number }): {
  lines: JournalLineInput[];
  description: string;
} {
  const memo = input.memo?.trim() || defaultMemo(input.kind, input.isPrive);
  const amount = input.amount;

  switch (input.kind) {
    case "cash_in": {
      if (!input.cashAccountId || !input.counterAccountId) {
        throw new ApiError(
          "VALIDATION_ERROR",
          "cashAccountId and counterAccountId are required for cash_in",
          400,
        );
      }
      return {
        description: memo,
        lines: [
          { accountId: input.cashAccountId, debit: amount, credit: 0 },
          { accountId: input.counterAccountId, debit: 0, credit: amount },
        ],
      };
    }
    case "cash_out": {
      if (!input.cashAccountId || !input.counterAccountId) {
        throw new ApiError(
          "VALIDATION_ERROR",
          "cashAccountId and counterAccountId are required for cash_out",
          400,
        );
      }
      return {
        description: memo,
        lines: [
          { accountId: input.counterAccountId, debit: amount, credit: 0 },
          { accountId: input.cashAccountId, debit: 0, credit: amount },
        ],
      };
    }
    case "transfer": {
      if (!input.cashAccountId || !input.counterAccountId) {
        throw new ApiError(
          "VALIDATION_ERROR",
          "cashAccountId (source) and counterAccountId (destination) are required for transfer",
          400,
        );
      }
      return {
        description: memo,
        lines: [
          { accountId: input.counterAccountId, debit: amount, credit: 0 },
          { accountId: input.cashAccountId, debit: 0, credit: amount },
        ],
      };
    }
    case "capital": {
      if (input.isPrive) {
        return {
          description: memo,
          lines: [
            {
              accountId: input.counterAccountId!,
              debit: amount,
              credit: 0,
            },
            { accountId: input.cashAccountId!, debit: 0, credit: amount },
          ],
        };
      }
      return {
        description: memo,
        lines: [
          { accountId: input.cashAccountId!, debit: amount, credit: 0 },
          {
            accountId: input.counterAccountId!,
            debit: 0,
            credit: amount,
          },
        ],
      };
    }
    default:
      throw new ApiError(
        "NOT_IMPLEMENTED",
        `Document kind '${input.kind}' is not implemented yet`,
        501,
      );
  }
}

function defaultMemo(kind: DocumentKind, isPrive?: boolean): string {
  switch (kind) {
    case "cash_in":
      return "Kas masuk";
    case "cash_out":
      return "Kas keluar";
    case "transfer":
      return "Transfer kas";
    case "capital":
      return isPrive ? "Prive" : "Setor modal";
    default:
      return "Transaksi";
  }
}

async function resolveAccounts(input: CreateDocumentInput): Promise<CreateDocumentInput> {
  const resolved = { ...input };

  if (resolved.kind === "capital") {
    resolved.cashAccountId =
      resolved.cashAccountId ?? (await getDefaultCashAccount(resolved.orgId));
    if (resolved.isPrive) {
      resolved.counterAccountId =
        resolved.counterAccountId ??
        (await getSystemAccountByCode(resolved.orgId, "3200"));
    } else {
      resolved.counterAccountId =
        resolved.counterAccountId ??
        (await getSystemAccountByCode(resolved.orgId, "3100"));
    }
    return resolved;
  }

  if (resolved.kind === "transfer") {
    if (!resolved.cashAccountId || !resolved.counterAccountId) {
      throw new ApiError(
        "VALIDATION_ERROR",
        "cashAccountId (source) and counterAccountId (destination) are required for transfer",
        400,
      );
    }
    const source = await getAccount(
      resolved.orgId,
      resolved.cashAccountId,
      "Source",
    );
    const dest = await getAccount(
      resolved.orgId,
      resolved.counterAccountId,
      "Destination",
    );
    if (!source.isCash || !dest.isCash) {
      throw new ApiError(
        "INVALID_ACCOUNT",
        "Transfer requires both accounts to be cash accounts",
        400,
      );
    }
    if (source.id === dest.id) {
      throw new ApiError(
        "INVALID_ACCOUNT",
        "Source and destination must differ",
        400,
      );
    }
    return resolved;
  }

  if (resolved.kind === "cash_in" || resolved.kind === "cash_out") {
    resolved.cashAccountId =
      resolved.cashAccountId ?? (await getDefaultCashAccount(resolved.orgId));
    if (!resolved.counterAccountId) {
      throw new ApiError(
        "VALIDATION_ERROR",
        "counterAccountId is required",
        400,
      );
    }
    const cash = await getAccount(
      resolved.orgId,
      resolved.cashAccountId,
      "Cash",
    );
    if (!cash.isCash) {
      throw new ApiError("INVALID_ACCOUNT", "Cash account must be a cash account", 400);
    }
  }

  return resolved;
}

const ARAP_KINDS = new Set<DocumentKind>([
  "invoice",
  "receipt",
  "loan_in",
  "loan_payment",
]);

export async function createDocument(
  input: CreateDocumentInput,
): Promise<{
  id: string;
  kind: string;
  status: string;
  date: string;
  amount: number;
  memo: string;
  openItemId?: string;
}> {
  if (ARAP_KINDS.has(input.kind)) {
    const arapInput = {
      orgId: input.orgId,
      userId: input.userId,
      amount: input.amount,
      memo: input.memo,
      date: input.date,
      dueDate: input.dueDate,
      contactId: input.contactId,
      openItemId: input.openItemId,
      documentId: input.documentId,
      cashAccountId: input.cashAccountId,
    };

    switch (input.kind) {
      case "invoice":
        return createInvoice(arapInput);
      case "receipt":
        return createReceipt(arapInput);
      case "loan_in":
        return createLoanIn(arapInput);
      case "loan_payment":
        return createLoanPayment(arapInput);
    }
  }

  if (!FULLY_IMPLEMENTED_KINDS.has(input.kind)) {
    throw new ApiError(
      "NOT_IMPLEMENTED",
      `Document kind '${input.kind}' is not implemented yet`,
      501,
    );
  }

  const amount = parseAmount(input.amount);
  const entryDate = parseDate(input.date);
  const resolved = await resolveAccounts(input);
  const documentType = KIND_TO_DOC_TYPE[resolved.kind];
  const { lines, description } = buildJournalLines({ ...resolved, amount });

  const result = await db.transaction(async (tx) => {
    const number = await nextDocumentNumber(resolved.orgId, documentType, tx);
    const dateStr = entryDate.toISOString().slice(0, 10);

    const [doc] = await tx
      .insert(documents)
      .values({
        orgId: resolved.orgId,
        type: documentType,
        number,
        contactId: resolved.contactId,
        date: dateStr,
        status: "posted",
        description,
        totalAmount: amount.toFixed(2),
        metadata: {
          kind: resolved.kind,
          isPrive: resolved.isPrive ?? false,
          cashAccountId: resolved.cashAccountId,
          counterAccountId: resolved.counterAccountId,
        },
      })
      .returning();

    const { entryId } = await postJournal({
      orgId: resolved.orgId,
      date: entryDate,
      description,
      lines,
      documentId: doc!.id,
      userId: resolved.userId,
      tx,
    });

    return { doc: doc!, entryId };
  });

  return {
    id: result.doc.id,
    kind: resolved.kind,
    status: result.doc.status,
    date: result.doc.date,
    amount,
    memo: result.doc.description ?? "",
  };
}

export function documentToApi(row: {
  id: string;
  type: string;
  status: string;
  date: string;
  totalAmount: string;
  description: string | null;
  metadata: Record<string, unknown> | null;
}): {
  id: string;
  kind: string;
  status: string;
  date: string;
  amount: number;
  memo: string;
} {
  const metadata = row.metadata ?? {};
  const kind =
    typeof metadata.kind === "string" ? metadata.kind : row.type;

  return {
    id: row.id,
    kind,
    status: row.status,
    date: row.date,
    amount: Math.round(Number(row.totalAmount)),
    memo: row.description ?? "",
  };
}

File: backend-go/internal/ledger/documents.go
package ledger

import (
	"context"
	"encoding/json"
	"math"
	"net/http"
	"strconv"
	"strings"
	"time"

	"github.com/icus/finbiz/backend-go/internal/platform"
	"github.com/jackc/pgx/v5"
	"github.com/jackc/pgx/v5/pgxpool"
)

// DocumentKind for cash documents.
type DocumentKind string

const (
	KindCashIn   DocumentKind = "cash_in"
	KindCashOut  DocumentKind = "cash_out"
	KindTransfer DocumentKind = "transfer"
	KindCapital  DocumentKind = "capital"
)

var kindToDocType = map[DocumentKind]DocumentType{
	KindCashIn:   DocReceipt,
	KindCashOut:  DocPayment,
	KindTransfer: DocJournal,
	KindCapital:  DocJournal,
}

var fullyImplemented = map[DocumentKind]struct{}{
	KindCashIn:   {},
	KindCashOut:  {},
	KindTransfer: {},
	KindCapital:  {},
}

// CreateDocumentInput is input for CreateDocument.
type CreateDocumentInput struct {
	OrgID            string
	UserID           string
	Kind             DocumentKind
	Amount           float64
	Memo             string
	CashAccountID    string
	CounterAccountID string
	IsPrive          bool
	Date             string
	ContactID        *string
}

// DocumentAPI is the API document shape.
type DocumentAPI struct {
	ID     string  `json:"id"`
	Kind   string  `json:"kind"`
	Status string  `json:"status"`
	Date   string  `json:"date"`
	Amount float64 `json:"amount"`
	Memo   string  `json:"memo"`
}

type accountInfo struct {
	ID     string
	Type   string
	IsCash bool
	Code   string
}

func parseDocumentAmount(amount float64) (float64, error) {
	rounded := math.Round(amount)
	if math.IsNaN(rounded) || math.IsInf(rounded, 0) || rounded <= 0 {
		return 0, platform.NewApiError(http.StatusBadRequest, "INVALID_AMOUNT", "Amount must be a positive integer")
	}
	return rounded, nil
}

func getAccount(ctx context.Context, db *pgxpool.Pool, orgID, accountID, label string) (*accountInfo, error) {
	var a accountInfo
	err := db.QueryRow(ctx, `
		SELECT id, type, is_cash, code FROM accounts WHERE id = $1 AND org_id = $2
	`, accountID, orgID).Scan(&a.ID, &a.Type, &a.IsCash, &a.Code)
	if err != nil {
		if err == pgx.ErrNoRows {
			return nil, platform.NewApiError(http.StatusNotFound, "ACCOUNT_NOT_FOUND", label+" account not found")
		}
		return nil, err
	}
	return &a, nil
}

func defaultMemo(kind DocumentKind, isPrive bool) string {
	switch kind {
	case KindCashIn:
		return "Kas masuk"
	case KindCashOut:
		return "Kas keluar"
	case KindTransfer:
		return "Transfer kas"
	case KindCapital:
		if isPrive {
			return "Prive"
		}
		return "Setor modal"
	default:
		return "Transaksi"
	}
}

func buildJournalLines(input CreateDocumentInput, amount float64) (lines []JournalLineInput, description string, err error) {
	memo := strings.TrimSpace(input.Memo)
	if memo == "" {
		memo = defaultMemo(input.Kind, input.IsPrive)
	}
	switch input.Kind {
	case KindCashIn:
		if input.CashAccountID == "" || input.CounterAccountID == "" {
			return nil, "", platform.NewApiError(http.StatusBadRequest, "VALIDATION_ERROR",
				"cashAccountId and counterAccountId are required for cash_in")
		}
		return []JournalLineInput{
			{AccountID: input.CashAccountID, Debit: amount},
			{AccountID: input.CounterAccountID, Credit: amount},
		}, memo, nil
	case KindCashOut:
		if input.CashAccountID == "" || input.CounterAccountID == "" {
			return nil, "", platform.NewApiError(http.StatusBadRequest, "VALIDATION_ERROR",
				"cashAccountId and counterAccountId are required for cash_out")
		}
		return []JournalLineInput{
			{AccountID: input.CounterAccountID, Debit: amount},
			{AccountID: input.CashAccountID, Credit: amount},
		}, memo, nil
	case KindTransfer:
		if input.CashAccountID == "" || input.CounterAccountID == "" {
			return nil, "", platform.NewApiError(http.StatusBadRequest, "VALIDATION_ERROR",
				"cashAccountId (source) and counterAccountId (destination) are required for transfer")
		}
		return []JournalLineInput{
			{AccountID: input.CounterAccountID, Debit: amount},
			{AccountID: input.CashAccountID, Credit: amount},
		}, memo, nil
	case KindCapital:
		if input.IsPrive {
			return []JournalLineInput{
				{AccountID: input.CounterAccountID, Debit: amount},
				{AccountID: input.CashAccountID, Credit: amount},
			}, memo, nil
		}
		return []JournalLineInput{
			{AccountID: input.CashAccountID, Debit: amount},
			{AccountID: input.CounterAccountID, Credit: amount},
		}, memo, nil
	default:
		return nil, "", platform.NewApiError(http.StatusNotImplemented, "NOT_IMPLEMENTED",
			"Document kind '"+string(input.Kind)+"' is not implemented yet")
	}
}

func resolveAccounts(ctx context.Context, db *pgxpool.Pool, input CreateDocumentInput) (CreateDocumentInput, error) {
	resolved := input

	if resolved.Kind == KindCapital {
		if resolved.CashAccountID == "" {
			id, err := GetDefaultCashAccount(ctx, db, resolved.OrgID)
			if err != nil {
				return resolved, err
			}
			resolved.CashAccountID = id
		}
		if resolved.CounterAccountID == "" {
			code := "3100"
			if resolved.IsPrive {
				code = "3200"
			}
			id, err := GetSystemAccountByCode(ctx, db, resolved.OrgID, code)
			if err != nil {
				return resolved, err
			}
			resolved.CounterAccountID = id
		}
		return resolved, nil
	}

	if resolved.Kind == KindTransfer {
		if resolved.CashAccountID == "" || resolved.CounterAccountID == "" {
			return resolved, platform.NewApiError(http.StatusBadRequest, "VALIDATION_ERROR",
				"cashAccountId (source) and counterAccountId (destination) are required for transfer")
		}
		source, err := getAccount(ctx, db, resolved.OrgID, resolved.CashAccountID, "Source")
		if err != nil {
			return resolved, err
		}
		dest, err := getAccount(ctx, db, resolved.OrgID, resolved.CounterAccountID, "Destination")
		if err != nil {
			return resolved, err
		}
		if !source.IsCash || !dest.IsCash {
			return resolved, platform.NewApiError(http.StatusBadRequest, "INVALID_ACCOUNT",
				"Transfer requires both accounts to be cash accounts")
		}
		if source.ID == dest.ID {
			return resolved, platform.NewApiError(http.StatusBadRequest, "INVALID_ACCOUNT",
				"Source and destination must differ")
		}
		return resolved, nil
	}

	if resolved.Kind == KindCashIn || resolved.Kind == KindCashOut {
		if resolved.CashAccountID == "" {
			id, err := GetDefaultCashAccount(ctx, db, resolved.OrgID)
			if err != nil {
				return resolved, err
			}
			resolved.CashAccountID = id
		}
		if resolved.CounterAccountID == "" {
			return resolved, platform.NewApiError(http.StatusBadRequest, "VALIDATION_ERROR", "counterAccountId is required")
		}
		cash, err := getAccount(ctx, db, resolved.OrgID, resolved.CashAccountID, "Cash")
		if err != nil {
			return resolved, err
		}
		if !cash.IsCash {
			return resolved, platform.NewApiError(http.StatusBadRequest, "INVALID_ACCOUNT", "Cash account must be a cash account")
		}
	}

	return resolved, nil
}

// CreateDocument creates a cash/capital document and posts its journal.
func CreateDocument(ctx context.Context, db *pgxpool.Pool, input CreateDocumentInput) (*DocumentAPI, error) {
	if _, ok := fullyImplemented[input.Kind]; !ok {
		return nil, platform.NewApiError(http.StatusNotImplemented, "NOT_IMPLEMENTED",
			"Document kind '"+string(input.Kind)+"' is not implemented yet")
	}

	amount, err := parseDocumentAmount(input.Amount)
	if err != nil {
		return nil, err
	}
	dateStr, err := documentDate(input.Date)
	if err != nil {
		return nil, err
	}
	entryDate, err := time.Parse("2006-01-02", dateStr)
	if err != nil {
		return nil, platform.NewApiError(http.StatusBadRequest, "INVALID_DATE", "Date must be YYYY-MM-DD")
	}

	resolved, err := resolveAccounts(ctx, db, input)
	if err != nil {
		return nil, err
	}
	documentType := kindToDocType[resolved.Kind]
	lines, description, err := buildJournalLines(resolved, amount)
	if err != nil {
		return nil, err
	}

	tx, err := db.Begin(ctx)
	if err != nil {
		return nil, err
	}
	defer tx.Rollback(ctx)

	number, err := NextDocumentNumber(ctx, tx, resolved.OrgID, documentType)
	if err != nil {
		return nil, err
	}

	meta, _ := json.Marshal(map[string]any{
		"kind":             string(resolved.Kind),
		"isPrive":          resolved.IsPrive,
		"cashAccountId":    resolved.CashAccountID,
		"counterAccountId": resolved.CounterAccountID,
	})

	var docID, status, docDate string
	var docDesc *string
	err = tx.QueryRow(ctx, `
		INSERT INTO documents (org_id, type, number, contact_id, date, status, description, total_amount, metadata)
		VALUES ($1, $2, $3, $4, $5, 'posted', $6, $7, $8)
		RETURNING id, status, date::text, description
	`, resolved.OrgID, string(documentType), number, resolved.ContactID, dateStr, description, formatAmount(amount), meta,
	).Scan(&docID, &status, &docDate, &docDesc)
	if err != nil {
		return nil, err
	}

	userID := resolved.UserID
	_, err = PostJournal(ctx, db, PostJournalInput{
		OrgID:       resolved.OrgID,
		Date:        entryDate,
		Description: description,
		Lines:       lines,
		DocumentID:  &docID,
		UserID:      &userID,
		Tx:          tx,
	})
	if err != nil {
		return nil, err
	}

	if err := tx.Commit(ctx); err != nil {
		return nil, err
	}

	memo := ""
	if docDesc != nil {
		memo = *docDesc
	}
	return &DocumentAPI{
		ID:     docID,
		Kind:   string(resolved.Kind),
		Status: status,
		Date:   docDate,
		Amount: amount,
		Memo:   memo,
	}, nil
}

func documentDate(date string) (string, error) {
	value := date
	if value == "" {
		value = time.Now().UTC().Format("2006-01-02")
	}
	if !dateRe.MatchString(value) {
		return "", platform.NewApiError(http.StatusBadRequest, "INVALID_DATE", "Date must be YYYY-MM-DD")
	}
	return value, nil
}

// DocumentToAPI maps a DB document row to API shape.
func DocumentToAPI(id, docType, status, date, totalAmount string, description *string, metadata []byte) DocumentAPI {
	kind := docType
	if len(metadata) > 0 {
		var m map[string]any
		if json.Unmarshal(metadata, &m) == nil {
			if k, ok := m["kind"].(string); ok && k != "" {
				kind = k
			}
		}
	}
	memo := ""
	if description != nil {
		memo = *description
	}
	f, _ := strconv.ParseFloat(totalAmount, 64)
	return DocumentAPI{
		ID:     id,
		Kind:   kind,
		Status: status,
		Date:   date,
		Amount: math.Round(f),
		Memo:   memo,
	}
}
